File: DSAMahidharSir/Graph/graphImplementation.c
#include<stdio.h>
#include<stdlib.h>
#include<stdbool.h>

typedef struct graph{
    int totalVertices;
    int totalEdges;
    int** adjList;
} Graph;

Graph* getGraph(int totalVertices){
    Graph* graph = (Graph*)malloc(sizeof(Graph));
    graph->totalVertices = totalVertices;
    graph->totalEdges = 0;
    graph->adjList = (int**)malloc(sizeof(int*)*graph->totalEdges);
    // printf("Graph created");
    return graph;
}

bool isVertex(Graph* graph, int u){
    return graph->totalVertices > u;
}

void addEdge(Graph* graph, int u, int v){
    if(!isVertex(graph,u) || !isVertex(graph,v)){
        printf("ValueError : Vertex doesn't exists\n");
        return;
    }
    else{
        graph->adjList = realloc(graph->adjList, sizeof(int*)*(++graph->totalEdges));
        graph->adjList[graph->totalEdges-1] = (int*)malloc(sizeof(int)*2);
        graph->adjList[graph->totalEdges-1][0] = u;
        graph->adjList[graph->totalEdges-1][1] = v;
    }
}

void removeEdge(Graph* graph, int u, int v){
    if(!isVertex(graph,u) || !isVertex(graph,v)){
        printf("ValueError : Vertex doesn't exists\n");
        return;
    }
    else{
        for (int i = 0; i < graph->totalEdges; i++)
        {
            if(graph->adjList[i][0] == u && graph->adjList[i][1] == v){
                graph->adjList[i][0] = graph->adjList[graph->totalEdges-1][0];
                graph->adjList[i][1] = graph->adjList[graph->totalEdges-1][1];
                graph->adjList = realloc(graph->adjList, sizeof(int*)*(--graph->totalEdges));
            }
        }   
    }
}

void freeGraph(Graph* graph){
    for (int i = 0; i < graph->totalEdges; i++){
        free(graph->adjList[i]);
    }
    free(graph->adjList);
    free(graph);
}

void printGraph(Graph* graph){
    if(!graph){
        return;
    }
    else{
        for(int i = 0; i < graph->totalEdges; i++){
            printf("[%d, %d]\n",graph->adjList[i][0], graph->adjList[i][1]);
        }
        printf("\n");
    }
}

int main(){
    Graph* graph = getGraph(6);
    addEdge(graph, 1, 2);
    addEdge(graph, 2, 3);
    addEdge(graph, 3, 1);
    addEdge(graph, 4, 1);
    addEdge(graph, 5, 1);
    addEdge(graph, 3, 4);

    removeEdge(graph, 2, 3);
    printGraph(graph);
    return 0;
}

File: DSAMahidharSir/Graph/graphImplementation.py
# class Vertex:
#     def __init__(self,data) -> None:
#         self.data = data 
#         self.next = None

# BFS = LEVEL ORDER TRAVERSAL
        
class Graph:
    # def __init__(self,num) -> None:
    #     self.vertices = num
    #     self.graph = [None] * self.vertices

    # def addEdges(self, src, dest):
    #     ver = Vertex(dest)
    #     ver.next = self.graph[src]
    #     self.graph[src] = ver

    #     ver = Vertex(src)
    #     ver.next = self.graph[dest]
    #     self.graph[dest] = ver

    # def printGraph(self):


    def __init__(self,vertex) -> None:
        self.vertex = vertex
        # self.matrix = [[None for _ in range(vertex)] for _ in range(vertex)]
        self.matrix=[[0]*self.vertex for _ in range(self.vertex)]
        for i in range(vertex):
            for j in range(vertex):
                if i == j:
                    self.matrix[i][j] = 0

    def isVertex(self,vertex):
        return vertex < self.vertex and 0 <= vertex

    def addEdge(self,u,v):
        if not self.isVertex(u) or not self.isVertex(v):
            print("Source or Destination not available")
            return
        else:
            if self.matrix[u][v] == 0:
                self.matrix[u][v] = 1
                self.matrix[v][u] = 1 
            else:
                raise(ValueError)
            
    def bfs(self,src):
        vis = []
        queue = [src]

        while queue:
            if queue[0] not in vis:
                vis.append(queue[0])
                print(queue[0], " ", end="")

            for node in self.adj_list[queue[0]]:
                if node not in vis:
                    queue.append(node)
            queue.pop(0)
            
    def removeEdge(self,u,v):
        if not self.isVertex(u) or not self.isVertex(v):
            print("Source or Destination not available")
            return
        else:
            if self.matrix[u][v] == 1:
                self.matrix[u][v] = 0
                self.matrix[v][u] = 0
            else:
                raise(ValueError)
            
    def printGraph(self):
        for vertex in self.matrix:
            print(*vertex)

    
            
gr = Graph(6)
gr.addEdge(0, 2)
gr.addEdge(2, 3)
gr.addEdge(3, 1)
gr.addEdge(4, 5)
gr.addEdge(3, 4)
gr.removeEdge(4, 5)
gr.bfs(0)
gr.printGraph()
    